### ISSUE TITLE
Question on copying startup to running datastore during image upgrade.

### ISSUE THREAD

**Issue by nehajoy** [sysrepo/sysrepo#2170]
Hi,

I have a yang model installed having a leaf of type uint16. The leaf is set to value 60 and this configuration is saved to startup. Then I upgrade the image with new version of the yang model where I have imposed a range constraint on the leaf to accept only values between 65 to 1000. As the startup configuration has a value 60 for this leaf, this value is copied from startup to running configuration without throwing any error during image upgrade. I am using sysrepo version 1.4.2. Is this expected or the validation for range should fail during bootup ?

Thanks in advance,
Neha

**Comment by michalvasko** [comment]
Hi,
I would expect you get an error during the "image upgrade", specifically when you update the YANG module (with `sysrepoctl -U` for example). Did you update the YANG module this way or in some other improper way?

**Comment by nehajoy** [comment]
Hi,

I installed the yang model again not just update.

Thanks,
Neha

**Comment by michalvasko** [comment]
In that case you have lost all the data of the module.

**Comment by nehajoy** [comment]
Basically before system upgrade I have copied the running configuration to startup configuration. In my device, during a system upgrade all yang models are installed again and then whatever configuration is present in startup it is copied to running configuration. So the stored values in startup now gets applied to running. But no validation failure occurs during writing to running datastore.

**Comment by michalvasko** [comment]
Like I said, you lose all the data when you uninstall the module including any stored in the `startup` datastore. So you are probably actually not copying anything on your final `startup` -> `running` copy, that is why you do not see any errors.

**Comment by nehajoy** [comment]
Startup configuration is stored in a persistent memory before system upgrade. During system upgrade the startup configuration files are re-instated from the persistent storage. This happens before starting sysrepo. The startup config is thus initialised before sysrepo starts and hence no data is lost. After sysrepo starts running, yang is installed and startup is copied to running datastore.

**Comment by michalvasko** [comment]
> Startup configuration is stored in a persistent memory before system upgrade. During system upgrade the startup configuration files are re-instated from the persistent storage.

Right, so you are tampering with *sysrepo* files. There is your problem.

**Comment by nehajoy** [comment]
I have the following yang model.

module test1 {
yang-version 1;
namespace "urn:ietf:params:xml:ns:yang:test1";
prefix test1;

organization
"XXX, Inc.";
contact
"XXX";
description
"Test1 Yang File";

revision 2020-04-03 {
description
"Initial Revision.";
}

leaf leaf1 {
type string;
default "string123";
description
"leaf1";
}

leaf leaf2 {
type uint16 {
  range "68..1500";
}
description
"leaf2";
}
}

I import the following xml to the running datastore with different values for leaf2. (No copying startup to running done here, just importing different values to running datastore using sysrepocfg import)

**Case 1 :**
```
thinkpalm@itp-wks0342:~$ cat test1.xml
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">68</leaf2>
```

```
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --import=/home/thinkpalm/test1.xml -f xml -m test1
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --export -f xml -m test1
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">68</leaf2>
```

**Case 2 :**
```
thinkpalm@itp-wks0342:~$ cat test1.xml
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">60</leaf2>
```

```
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --import=/home/thinkpalm/test1.xml -f xml -m test1
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --export -f xml -m test1
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">60</leaf2>
```

**Case 3 :**
```
thinkpalm@itp-wks0342:~$ cat test1.xml
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">1501</leaf2>
```

```
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --import=/home/thinkpalm/test1.xml -f xml -m test1
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --export -f xml -m test1
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">1501</leaf2>
```

**Case 4 :**
```
thinkpalm@itp-wks0342:~$ cat test1.xml
<leaf2 xmlns="urn:ietf:params:xml:ns:yang:test1">65536</leaf2>
```

```
thinkpalm@itp-wks0342:~/sysrepo/examples$ sudo sysrepocfg --import=/home/thinkpalm/test1.xml -f xml -m test1
sysrepocfg error: libyang: Invalid value "65536" in "leaf2" element.
sysrepocfg error: Data parsing failed
```


According to my understanding, out of range values (values less than 68 and greater than 1500) should be rejected. But as you can see only values out of the scope of uint16 type is rejected here. Can you please explain why the range constraint is not applied here ?

Thanks,
Neha

**Comment by michalvasko** [comment]
Hi,
I see. This was an undesirable consequence of how validation was performed in the version you are using. If you update *libyang* (you will probably need to update *sysrepo* as well), it should work as expected. Generally, you should always use at least the latest release.

**Comment by nehajoy** [comment]
Yes, I tested with latest version and issue is not seen. For our product, we are not currently migrating to latest version. Can I get the patch where this issue is fixed ?

Thanks,
Neha

**Comment by michalvasko** [comment]
It should be [this commit](https://github.com/CESNET/libyang/commit/0abb1bc3f95876f82625e8a0aadd973ee4a58db5). Even though I do not think it will happen in this case, when selecting patches, you always risk that the change depends on some other one.
